src/peft/tuners/lora/config.py

```diff
@@ -130,3 +130,13 @@ def __post_init__(self):
         # if target_modules is a regex expression, then layers_pattern should be None
         if isinstance(self.target_modules, str) and self.layers_pattern is not None:
             raise ValueError("`layers_pattern` cannot be used when `target_modules` is a str.")
+
+
+class MegatronConfig:
```

Comment:
> > Due to the particularity of megatron creating a distributed layer, the required megatron information needs to be injected before executing Lora
> 
> I still don't understand this fully. Ideally, we should find a solution that allows these config options to just live on `LoraConfig`.
> 
> What I can tell is that in `tp_layer.py`, the `LoraParallelLinear` inherits from `MegatronConfig.tensor_parallel`, which is probably where the issue is. It seems to me that there are multiple backends and you need the backend to be determined before the `LoraParallelLinear` class is created, is that right?
> 
> One solution that comes to mind is that we create one class for each possible backend. The code could look somewhat like this:
> 
> ```python
> class _LoraParallelLinearBase:
>     def __init__(...)  # same code as LoraParallelLinear, but use self.backend for backend-specific stuff
> 
> # assume we have the backends foo and bar:
> 
> class LoraParallelLinearFoo(_LoraParallelLinearBase, foo.ColumnParallelLinear, foo.RowParallelLinear, LoraParalleLayer):
>     def __init__(self, *args, **kwargs)
>         self.backend = foo
>         super().__init__(*args, **kwargs)
> 
> class LoraParallelLinearBar(_LoraParallelLinearBase, bar.ColumnParallelLinear, bar.RowParallelLinear, LoraParalleLayer):
>     def __init__(self, *args, **kwargs)
>         self.backend = bar
>         super().__init__(*args, **kwargs)
> ```
> 
> Probably a few more changes need to be made, but I hop this gets the idea across.
> 
> If this doesn't work for some reason, I'm sure we can find other ways, maybe by creating the class at runtime. As is, however, I think it's difficult for the user to make use of the feature.


Okay, I'll try to think of a way without explicit injection.

Let me first explain why we need to design it this way:
Because currently when people use the two warehouses Megatron-LM or Megatron-DeepSpeed, most of them will not install them into the environment. Instead, they will make some magic changes locally and then start training directly in the source code directory. (like us)

So I plan to change to using the native megatron class by default, so that explicit injection is not needed. But I plan to keep the set injection interface for people like us who have special needs.

Reply:
> > Due to the particularity of megatron creating a distributed layer, the required megatron information needs to be injected before executing Lora
> 
> I still don't understand this fully. Ideally, we should find a solution that allows these config options to just live on `LoraConfig`.
> 
> What I can tell is that in `tp_layer.py`, the `LoraParallelLinear` inherits from `MegatronConfig.tensor_parallel`, which is probably where the issue is. It seems to me that there are multiple backends and you need the backend to be determined before the `LoraParallelLinear` class is created, is that right?
> 
> One solution that comes to mind is that we create one class for each possible backend. The code could look somewhat like this:
> 
> ```python
> class _LoraParallelLinearBase:
>     def __init__(...)  # same code as LoraParallelLinear, but use self.backend for backend-specific stuff
> 
> # assume we have the backends foo and bar:
> 
> class LoraParallelLinearFoo(_LoraParallelLinearBase, foo.ColumnParallelLinear, foo.RowParallelLinear, LoraParalleLayer):
>     def __init__(self, *args, **kwargs)
>         self.backend = foo
>         super().__init__(*args, **kwargs)
> 
> class LoraParallelLinearBar(_LoraParallelLinearBase, bar.ColumnParallelLinear, bar.RowParallelLinear, LoraParalleLayer):
>     def __init__(self, *args, **kwargs)
>         self.backend = bar
>         super().__init__(*args, **kwargs)
> ```
> 
> Probably a few more changes need to be made, but I hop this gets the idea across.
> 
> If this doesn't work for some reason, I'm sure we can find other ways, maybe by creating the class at runtime. As is, however, I think it's difficult for the user to make use of the feature.

After thinking about it, I plan to put the config used when creating the tp distributed layer into LoraConfig. In this case, I might as well put the reference to tensor_parallel into LoraConfig~:laughing:

Reply:
> > Due to the particularity of megatron creating a distributed layer, the required megatron information needs to be injected before executing Lora
> 
> I still don't understand this fully. Ideally, we should find a solution that allows these config options to just live on `LoraConfig`.
> 
> What I can tell is that in `tp_layer.py`, the `LoraParallelLinear` inherits from `MegatronConfig.tensor_parallel`, which is probably where the issue is. It seems to me that there are multiple backends and you need the backend to be determined before the `LoraParallelLinear` class is created, is that right?
> 
> One solution that comes to mind is that we create one class for each possible backend. The code could look somewhat like this:
> 
> ```python
> class _LoraParallelLinearBase:
>     def __init__(...)  # same code as LoraParallelLinear, but use self.backend for backend-specific stuff
> 
> # assume we have the backends foo and bar:
> 
> class LoraParallelLinearFoo(_LoraParallelLinearBase, foo.ColumnParallelLinear, foo.RowParallelLinear, LoraParalleLayer):
>     def __init__(self, *args, **kwargs)
>         self.backend = foo
>         super().__init__(*args, **kwargs)
> 
> class LoraParallelLinearBar(_LoraParallelLinearBase, bar.ColumnParallelLinear, bar.RowParallelLinear, LoraParalleLayer):
>     def __init__(self, *args, **kwargs)
>         self.backend = bar
>         super().__init__(*args, **kwargs)
> ```
> 
> Probably a few more changes need to be made, but I hop this gets the idea across.
> 
> If this doesn't work for some reason, I'm sure we can find other ways, maybe by creating the class at runtime. As is, however, I think it's difficult for the user to make use of the feature.

@BenjaminBossan I refactored the code according to the idea of ​​not explicitly injecting, please help me review it again. 
Now, we can use it like this:
```python
from megatron.arguments import core_transformer_config_from_args
from megatron import get_args
from megatron.core import tensor_parallel
from peft import LoraConfig, get_peft_model

config = core_transformer_config_from_args(get_args())

lora_config = LoraConfig(
    r=16,
    lora_alpha=32,
    target_modules=['query_key_value', 'dense', 'dense_h_to_4h', 'dense_4h_to_h'],
    lora_dropout=0.0,
    bias="none",
    megatron_config=config,
    tensor_parallel=tensor_parallel,
)
model = get_peft_model(model, lora_config)
model.print_trainable_parameters()
```


By the way, maybe I will resolve the conflicts and unit test code tomorrow.